ethereum/contracts/state-transition/StateTransition.sol

```diff
@@ -0,0 +1,212 @@
+        require(IBridgehub(bridgehub).getChainStateTransition(_chainId) == address(this), "PRegsitry 2");
+
+        // check input
+        bytes32 cutHash = keccak256(abi.encode(_diamondCut));
+        require(cutHash == stateTransitionStorage.cutHash, "PRegistry 3");
+
+        // construct init data
+        bytes memory initData;
+        bytes memory copiedData = _diamondCut.initCalldata[228:];
+        initData = bytes.concat(
+            IDiamondInit.initialize.selector,
+            bytes32(_chainId),
+            bytes32(uint256(uint160(address(bridgehub)))),
+            bytes32(uint256(uint160(address(this)))),
+            bytes32(uint256(stateTransitionStorage.protocolVersion)),
+            bytes32(uint256(uint160(_governor))),
+            bytes32(uint256(uint160(_governor))),
+            bytes32(stateTransitionStorage.storedBatchZero),
+            copiedData
+        );
+        Diamond.DiamondCutData memory cutData = _diamondCut;
+        cutData.initCalldata = initData;
+
+        // deploy stateTransitionChainContract
+        DiamondProxy stateTransitionChainContract = new DiamondProxy(block.chainid, cutData);
+
+        // save data
+        address stateTransitionChainAddress = address(stateTransitionChainContract);
+
+        stateTransitionStorage.stateTransitionChainContract[_chainId] = stateTransitionChainAddress;
+        stateTransitionStorage.chainNumberToContract[stateTransitionStorage.totalChains] = stateTransitionChainAddress;
+        ++stateTransitionStorage.totalChains;
+
+        IBridgehub(bridgehub).setStateTransitionChainContract(_chainId, stateTransitionChainAddress);
+
+        // set chainId in VM
+        _setChainIdUpgrade(_chainId, stateTransitionChainAddress);
+
+        emit StateTransitionNewChain(_chainId, stateTransitionChainAddress);
+    }
+
+    function setUpgradeDiamondCut(
+        Diamond.DiamondCutData calldata _cutData,
+        uint256 _protocolVersion
+    ) external onlyGovernor {
+        stateTransitionStorage.upgradeCutHash[_protocolVersion] = keccak256(abi.encode(_cutData));
+    }
+
+    function upgradeChain(
+        uint256 _chainId,
+        uint256 _protocolVersion,
+        Diamond.DiamondCutData calldata _cutData
+    ) external onlyChainGovernor(_chainId) {
+        bytes32 cutHash = keccak256(abi.encode(_cutData));
+        require(cutHash == stateTransitionStorage.upgradeCutHash[_protocolVersion], "r25");
+
+        IStateTransitionChain stateTransitionChainContract = IStateTransitionChain(
+            stateTransitionStorage.stateTransitionChainContract[_chainId]
+        );
+        stateTransitionChainContract.executeUpgrade(_cutData, _protocolVersion);
+    }
+
+    function specialUpgradeChain(
+        uint256 _chainId,
+        uint256 _protocolVersion,
+        Diamond.DiamondCutData calldata _cutData
+    ) external onlyGovernor {
+        IStateTransitionChain stateTransitionChainContract = IStateTransitionChain(
```

Comment:
Do we want to have `         require(cutHash == stateTransitionStorage.upgradeCutHash[_protocolVersion], "r25");
` here too? To prevent accidental mistakes of the governor

Reply:
Ok, then I will just delete this function, because then it is the same as the one right above.

Reply:
Then there is no point to this function, as it would be the same one as before. I removed the function